libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,324 @@
+
+		// For events kind() == Event, so we have an extra check here
+		if (funType && funType->kind() == FunctionType::Kind::Internal)
+			processFunction(callable->resolveVirtual(*m_contract), _identifier.annotation().calledDirectly);
+	}
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(NewExpression const& _newExpression)
+{
+	if (ContractType const* contractType = dynamic_cast<ContractType const*>(_newExpression.typeName().annotation().type))
+		m_graph->createdContracts.emplace(&contractType->contractDefinition());
+
+	return true;
+}
+
+void FunctionCallGraphBuilder::endVisit(MemberAccess const& _memberAccess)
+{
+	auto functionType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type);
+	auto functionDef = dynamic_cast<FunctionDefinition const*>(_memberAccess.annotation().referencedDeclaration);
+	if (!functionType || !functionDef || functionType->kind() != FunctionType::Kind::Internal)
+		return;
+
+	// Super functions
+	if (*_memberAccess.annotation().requiredLookup == VirtualLookup::Super)
+	{
+		if (auto const* typeType = dynamic_cast<TypeType const*>(_memberAccess.expression().annotation().type))
+			if (auto const contractType = dynamic_cast<ContractType const*>(typeType->actualType()))
+			{
+				solAssert(contractType->isSuper(), "");
+				functionDef =
+					&functionDef->resolveVirtual(
+						*m_contract,
+						contractType->contractDefinition().superContract(*m_contract)
+					);
+			}
+	}
+	else
+		solAssert(*_memberAccess.annotation().requiredLookup == VirtualLookup::Static, "");
+
+	processFunction(*functionDef, _memberAccess.annotation().calledDirectly);
+	return;
+}
+
+void FunctionCallGraphBuilder::endVisit(ModifierInvocation const& _modifierInvocation)
+{
+	VirtualLookup const& requiredLookup = *_modifierInvocation.name().annotation().requiredLookup;
+
+	if (auto const* modifier = dynamic_cast<ModifierDefinition const*>(_modifierInvocation.name().annotation().referencedDeclaration))
+	{
+		if (requiredLookup == VirtualLookup::Virtual)
+			modifier = &modifier->resolveVirtual(*m_contract);
+		else
+			solAssert(requiredLookup == VirtualLookup::Static, "");
+
+		processFunction(*modifier);
+	}
+}
+
+void FunctionCallGraphBuilder::visitCallable(CallableDeclaration const* _callable)
+{
+	solAssert(!m_graph->edges.count(_callable), "");
+
+	std::optional<Node> previousNode = m_currentNode;
+	m_currentNode = _callable;
+
+	_callable->accept(*this);
+
+	m_currentNode = previousNode;
+}
+
+bool FunctionCallGraphBuilder::add(Node _caller, Node _callee)
+{
+	return m_graph->edges[_caller].insert(_callee).second;
+}
+
+void FunctionCallGraphBuilder::processFunction(CallableDeclaration const& _callable, bool _calledDirectly)
+{
+	solAssert(m_currentNode.has_value(), "");
+
+	if (_calledDirectly)
+		add(*m_currentNode, &_callable);
+	else
+		add(m_currentDispatch, &_callable);
+
+	if (!m_graph->edges.count(&_callable))
+		visitCallable(&_callable);
```

Comment:
I would feel safer if we just added this to a queue instead of visiting it right away.

Reply:
Good point. Right now it's recursive and we might run out of stack space on some very long call chains. I'll change that.